### PR TITLE
Database optimization

### PR DESCRIPTION
**Pull request by SubJunk**

### CODE REVIEW

File: src/main/java/net/pms/image/ImagesUtil.java [hunk 1]
```diff
@@ -1479,7 +1479,7 @@ protected static Image transcodeImage(
 		Thumbnails.of(bufferedImage)
 			.scale(1.0d)
 			.outputFormat(outputFormat.toString())
-			.outputQuality(1.0f)
+			.outputQuality(0.6f)
```
Comment (Nadahar) [inline src/main/java/net/pms/image/ImagesUtil.java]: As I already said in #1527 this is effectively a bug.

Comment (SubJunk) [reply src/main/java/net/pms/image/ImagesUtil.java]: Is this better? https://github.com/UniversalMediaServer/UniversalMediaServer/pull/1528/commits/c95f6b3bcc65dee219f526426fbb2657ccc96a84

File: src/main/java/net/pms/dlna/DLNAMediaDatabase.java [hunk 2]
```diff
@@ -1252,6 +1256,62 @@ public synchronized void cleanup() {
...
+			 * Removes entries that are not referenced by any rows in the FILES table.
+			 */
+			ps = conn.prepareStatement(
+				"DELETE FROM SUBTRACKS " +
+				"WHERE NOT EXISTS (" +
+					"SELECT ID FROM FILES " +
+					"WHERE FILES.ID = SUBTRACKS.FILEID" +
+				");"
+			);
+			rs = ps.executeQuery();
```
Comment (atamariya) [inline src/main/java/net/pms/dlna/DLNAMediaDatabase.java]: How about adding DELETE CASCADE constraint instead to child tables?

Comment (SubJunk) [reply src/main/java/net/pms/dlna/DLNAMediaDatabase.java]: That's a good idea, thanks. That should work for all of them except the THUMBNAILS table

Comment (SubJunk) [reply src/main/java/net/pms/dlna/DLNAMediaDatabase.java]: That's done now, not fully tested but is that what you meant?

Comment (atamariya) [reply src/main/java/net/pms/dlna/DLNAMediaDatabase.java]: Yes, the constraints look fine.

**Comment by valib** [comment]
@SubJunk there is a problem with the musicbrainz.org restriction to make only limited number of requests in the limited time othervise the server returns status code 503. The same apply for the @Nadahar implementation  https://github.com/DigitalMediaServer/DigitalMediaServer/commits/DeduplicateAudioThumbnails. This problem was reported by @Nadahar some time ago. You can reproduce this bug using the web interface. When this happen and exception is made and for the thumbnail is not used the general thumbnail but the last thumbnail made for different media.

**Comment by Nadahar** [comment]
@valib 
> When this happen and exception is made and for the thumbnail is not used the general thumbnail but the last thumbnail made for different media.

If that's the case, it's a bug. I don't quite understand how this can be the case though, it seems very strange if it should use a thumbnail for a different media. Are you sure that's not just a bug with the web interface?

**Comment by valib** [comment]
@Nadahar the web interface is making max 100 requests in paralel which set in the PMSConfiguration  `WEB_MAX_THREADS = 100` but is not the case of the standard renderers whose are obviously making requests only for number which is shown at the screen.

**Comment by Nadahar** [comment]
@valib I wasn't thinking of the `503` error, but that the thumbnail for one media is shown for another. You can get the `503` error with normal renderers too, if you just browse a folder with many audio tracks in the, or browse multiple folders quickly. I just can't understand how a thumbnail for one media can be displayed for another, I would think that it would either show no thumbnail at all or a generic thumbnail.

**Comment by Nadahar** [comment]
@valib Forget it, I misunderstood your original text. I thought you were talking about my implementation, but I see from the log file that you weren't. I haven't looked at the implementation details for the other, so I don't know if that might happen or not.

**Comment by valib** [comment]
@Nadahar I am refering also to your DMS DeduplicateAudioThumbnails branch which has the same result..

**Comment by Nadahar** [comment]
@valib I've been unable to confirm the bug you describe. I did however find another "problem" when testing this that causes a "similar" issue. I did get a lot of thumbnails that didn't belong to the correct files for those that couldn't be looked up. When looking into it, it turned out that it was because it falls back to "folder thumbnails" when no thumbnail can be found for that particular media. There is quite a lot of "mess" in the folder I used for testing, so it turns out it misidentified some images in the folder for being "folder thumbnails". My solution to that problem is: https://github.com/DigitalMediaServer/DigitalMediaServer/commit/16c3ccc5e8bf2bbd55e93161942b34290c039e81

In addition I had to delete a couple of files that actually had matching "folder thumbnail" names. After doing this, generic thumbnails are shown for those that can't be fetched.

Can you confirm if this fixes your issue too? As this isn't really related to neither my or @SubJunk's implementation, it's likely that this will cause the same issue when you test both branches.

**Comment by valib** [comment]
@Nadahar I was wrong to inform that the last thumbnail is repeated. I had also the mess in my testing folder and there were the `folder.jpg` which was shown when the cover was not retrieved from the database because of the status code 503 returned. This is the main problem with the web interface and maybe with other renderers too when we not keep the MusicBrainz regulation `For "anonymous" user-agents we allow through (on average) 50 requests per second, and decline (http 503) the rest.`

**Comment by Nadahar** [comment]
@valib I'm aware of the query rate limitation on musicbrainz.org. It wasn't there when I made this. To be able to control the query rate, a different solution would have to be implemented, as now the query is made when the thumbnail is requested. The results are cached in the database though, so it should ideally just query once per media file. But, one query can be up to 4 actual requests depending on if it finds something, so it's a bit hard to follow the restrictions they have. If you read below the text you quoted, there's also a limit per IP address - which is one request per second. That is very little if I read this correctly.

**Comment by Nadahar** [comment]
@SubJunk I think it makes sense to look at how you're going to store the data to decide when to use joins. Since both the information from `FILES` and the thumbnail goes into `DLNAMediaInfo` I think it's correct to do a join there, especially since it doesn't seem that you will have any issues as long as you do a left outer join. Joining the other tables where there can be multiple entries per FILES entry is more of a challenge, and since they are used to create instances of different types, I can't see any point in using join there. First H2 would have to join all that information, and then our code would have to split it again, solving repeated information for the `FILES` table as you would have to do an outer join.

**Comment by SubJunk** [comment]
I made the change from your review @valib

**Comment by SubJunk** [comment]
@valib yes I think it would be a good idea to contact them about raising the limit for our user-agent

**Comment by valib** [comment]
@SubJunk there is a bug at the line https://github.com/UniversalMediaServer/UniversalMediaServer/blob/04aebdb197d5cb53dbc79387c353bb37bba20aec/src/main/java/net/pms/dlna/DLNAMediaDatabase.java#L536

**Comment by SubJunk** [comment]
Good catch, that's fixed now. Thanks
